components/script/dom/bindings/structuredclone.rs

```diff
@@ -5,15 +5,101 @@
 //! This module implements structured cloning, as defined by [HTML]
 //! (https://html.spec.whatwg.org/multipage/#safe-passing-of-structured-data).
 
+use dom::bindings::conversions::get_dom_class;
 use dom::bindings::error::{Error, Fallible};
 use dom::globalscope::GlobalScope;
-use js::jsapi::{HandleValue, MutableHandleValue};
-use js::jsapi::{JSContext, JS_ReadStructuredClone, JS_STRUCTURED_CLONE_VERSION};
-use js::jsapi::{JS_ClearPendingException, JS_WriteStructuredClone};
+use js::glue::UnwrapObject;
+use js::jsapi::{Handle, HandleObject, HandleValue, MutableHandleValue};
+use js::jsapi::{Heap, JSContext};
+use js::jsapi::{JSStructuredCloneCallbacks, JSStructuredCloneReader, JSStructuredCloneWriter};
+use js::jsapi::{JS_ClearPendingException, JSObject, JS_ReadStructuredClone};
+use js::jsapi::{JS_STRUCTURED_CLONE_VERSION, JS_WriteStructuredClone};
+use js::jsapi::{MutableHandleObject, TransferableOwnership};
 use libc::size_t;
+use std::os::raw;
 use std::ptr;
 use std::slice;
 
+
+pub enum StructuredCloneTags {
+    DomBlob = 0xffff8001
+}
+
+
+#[allow(dead_code)]
+unsafe extern "C" fn read_callback(_cx: *mut JSContext,
+                                   _r: *mut JSStructuredCloneReader,
+                                   tag: u32,
+                                   _data: u32,
+                                   _closure: *mut raw::c_void) -> *mut JSObject {
+    match tag {
+        tag if tag == StructuredCloneTags::DomBlob as u32 => {
```

Comment:
this also seems like a terrible hack, any suggestions on a better way to approach this?